apiserver/internal/handlers/resource/resource.go

```diff
@@ -0,0 +1,408 @@
+		if err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("returning error to user: %v", err)
+			}
+			return
+		}
+		if err := sendStatusAndJSON(resp, http.StatusOK, &response); err != nil {
+			h.logger.Errorf("sending response: %v", err)
+		}
+	default:
+		if err := sendError(resp, jujuerrors.MethodNotAllowedf("unsupported method: %q", req.Method), h.logger); err != nil {
+			h.logger.Errorf("returning error to user: %v", err)
+		}
+	}
+}
+
+func (h *ResourceHandler) download(service ResourceService, req *http.Request) (io.ReadCloser, int64, error) {
+	query := req.URL.Query()
+	application := query.Get(":application")
+	name := query.Get(":resource")
+
+	uuid, err := service.GetResourceUUIDByApplicationAndResourceName(req.Context(), application, name)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s", name, application)
+	} else if err != nil {
+		return nil, 0, fmt.Errorf("getting resource uuid: %w", err)
+	}
+
+	res, reader, err := service.OpenResource(req.Context(), uuid)
+	if errors.Is(err, resourceerrors.ResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s", name, application)
+	} else if errors.Is(err, resourceerrors.StoredResourceNotFound) {
+		return nil, 0, jujuerrors.NotFoundf("resource %s of application %s has no blob downloaded on controller", name, application)
+	} else if err != nil {
+		return nil, 0, errors.Errorf("opening resource %s for application %s: %w", name, application, err)
+	}
+	return reader, res.Size, err
+}
+
+func (h *ResourceHandler) upload(resourceService ResourceService, req *http.Request, username string) (_ *params.UploadResult, err error) {
+
+	reader, uploaded, err := h.getUploadedResource(resourceService, req)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	defer func() {
+		if err != nil {
+			reader.Close()
+		}
+	}()
+
+	err = resourceService.StoreResourceAndIncrementCharmModifiedVersion(
+		req.Context(),
+		resource.StoreResourceArgs{
+			ResourceUUID:    uploaded.uuid,
+			Reader:          reader,
+			RetrievedBy:     username,
+			RetrievedByType: coreresource.User,
+			Size:            uploaded.size,
+			Fingerprint:     uploaded.fingerprint,
+			Origin:          charmresource.OriginUpload,
+			Revision:        -1,
+		},
+	)
+	if err != nil {
+		return nil, errors.Errorf("storing resource %s of application %s: %w", uploaded.resourceName, uploaded.applicationName, err)
+	}
+
+	res, err := resourceService.GetResource(req.Context(), uploaded.uuid)
+	if err != nil {
+		return nil, errors.Errorf("getting uploaded resource details: %w", err)
+	}
+
+	return &params.UploadResult{
+		Resource: domainResource2API(res),
+	}, nil
+}
+
+// DomainResource2API converts a [domainresource.Resource] into
+// a [params.Resource] struct.
+func domainResource2API(res coreresource.Resource) params.Resource {
```

Comment:
It's not even domain resource, just call it `encodeResouce`. Also, I can't stand the number 2 in the method name, spell it out.

Reply:
Fair, this changed in a rebase and I missed it.